README.md
Logentries agent
================

[![License](https://img.shields.io/badge/license-MIT-blue.svg?style=flat)](https://github.com/mdp/rotp/blob/master/LICENSE)

A command line utility for a convenient access to Logentries logging
infrastructure.

  * [How to use](#how-to-use)
  * [Repositories](#repositories)
  * [Configuration file](#configuration-file)
  * [Follow log files through server-side configuration](#follow-log-files-through-server-side-configuration)
  * [Follow log files through your configuration file](#follow-log-files-through-your-configuration-file)
  * [Using local configuration only](#using-local-configuration-only)
  * [List IP addresses the agent uses](#list-ip-addresses-the-agent-uses)
  * [Follow logs that change their names](#follow-logs-that-change-their-names)
  * [Follow a log that appears across multiple directories](#follow-a-log-that-appears-across-multiple-directories)
  * [Save state of file positions](#state-file)
  * [Manipulate your data in transit](#manipulate-your-data-in-transit)
  * [Format output entries](#format-output-entries)
  * [Filtering file names](#filtering-file-names)
  * [Multiline log entries](#multiline-log-entries)
  * [System metrics (beta)](#system-metrics-beta)
    * [CPU](#cpu)
    * [VCPU](#vcpu)
    * [Memory](#memory)
    * [Swap](#swap)
    * [Network](#network)
    * [Disk IO](#disk-io)
    * [Disk space](#disk-space)
    * [Processes](#processes)
  * [Deployment best practices](#deployment-best-practices)
  * [Linux Agent Installation](#linux-agent-le-agent-installation)
  * [Contact Support](#contact-support)


How to use
----------

	usage: le COMMAND [ARGS]

	Where COMMAND is one of:
	init      Write local configuration file
	reinit    As init but does not reset undefined parameters
	register  Register this host
	--name=  name of the host
	--hostname=  hostname of the host
	whoami    Displays settings for this host
	monitor   Monitor this host
	follow <filename>  Follow the given log
	--name=  name of the log
	--type=  type of the log
	--multilog option used with directory wildcard (restricted behaviour - see below for details)
	followed <filename>  Check if the file is followed
	clean     Removes configuration file
	ls        List internal filesystem and settings: <path>
	ls ips    List IP addresses used by the agent
	rm        Remove entity: <path>
	pull      Pull log file: <path> <when> <filter> <limit>

	Where ARGS are:
	--help            show usage help and exit
	--version         display version number and exit
	--account-key=    set account key and exit
	--host-key=       set local host key and exit, generate key if key is empty
	--no-timestamps   no timestamps in agent reportings
	--force           force given operation
	--datahub         send logs to the specified data hub address
	                  the format is address:port with port being optional
	--suppress-ssl    do not use SSL with API server
	--yes	          always respond yes
	--pull-server-side-config=False do not use server-side config for following files


Repositories
------------

For Debian/Ubuntu systems include this line in `/etc/apt/sources.list.d/logentries.list`:

	deb http://rep.logentries.com/ XXX main

Replace `XXX` with the name of your system, i.e. one of wheezy, jessie,
lucid, precise, quantal, saucy, trusty, utopic, vivid. You also need to add
Logentries release key:

	apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv-keys A5270289C43C79AD

(Keyservers are not always reliable. In automated scripts do the following.)

	apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv-keys A5270289C43C79AD \
	|| apt-key adv --keyserver hkp://pgp.mit.com:80 --recv-keys A5270289C43C79AD

Then run `apt-get update` and `apt-get install logentries`. If you want to run
the agent as daemon, install it via `apt-get install logentries-daemon`.

For rpm-based systems RH, CentOS, Fedora, add this in `/etc/yum.repos.d/logentries.repo`

	[logentries]
	name=Logentries repo
	enabled=1
	metadata_expire=1d
	baseurl=http://rep.logentries.com/XXX/\$basearch
	gpgkey=http://rep.logentries.com/RPM-GPG-KEY-logentries

Replace `XXX` with the name of your system, i.e. one of fedora18, fedora19,
fedora20, fedora21, rh5, rh6, amazonlatest, centos5, centos6. Then run `yum
update` and `yum install logentries`. If you want to run the agent as daemon,
install it via `yum install logentries-daemon`.


Configuration file
------------------

The agent stores configuration in `~/.le/config` for ordinary users and in
`/etc/le/config` for root (daemon). It is created with `init` or `reinit`
commands and can be created or modified manually.

The agent supports loading multiple configuration files from a configuration
directory. By default the configuration directory is `~/.le/conf.d/` for
ordinary users and `/etc/le/conf.d` for root (daemon). Only files with `.conf`
extension are recognized as configuration files.

The structure of a configuration file follows standard similar to what you
find in `.git/config` or Windows INI files. For example:

	[Main]
	user-key = e720a1e8-a7d5-4f8b-8879-854e51c9290d
	agent-key = 428b888a-29ab-4079-99ec-9cb7aa2ffea7

	[cassandra]
	metrics-process = org.apache.cassandra.service.CassandraDaemon
	path = /var/log/cassandra/system.log
	token = a846bd59-a674-4088-b9fd-e72da1df5946

Main section `[Main]` contains agent-wide general configuration. Any other
section defines per-application settings such as log filenames and metrics.

In the main section, `user-key` (account key) which identifies account, and
`agent-key` which identifies host (host key).

Note the `monitor` command requires both `user-key` and `agent-key` defined.

### User-specified configuration directory

Additionaly to the default configuration directory, the agent allows to specify
a secondary configuration directory. Just specify the `include` parameter in
the main section as follows:

	[Main]
	user-key = e720a1e8-a7d5-4f8b-8879-854e51c9290d
	include = /opt/configurations/logentries/conf.d

Again, the agent recognizes all files with `.conf` extension.


Follow log files through server-side configuration
--------------------------------------------------

After registering the host (via `register` command or specifying `agent-key` in
configuration) you can add a file to follow via `follow` command:

	sudo le follow /srv/log/cassandra/system.out [--name Cassandra]

You can repeat the command for additional logs. The agent creates a new log
entry in Logentries under the host specified. It will also enable the file to
be followed by the agent.

Note `--name` is optional to specify log name as it will appear in UI and log
listing. If not specified, plain file name is used.

You need to restart the agent to pick up the new configuration:

	sudo service logentries restart


Follow log files through your configuration file
------------------------------------------------

Apart from server-side configuration you can configure log files to be followed
locally. Locally configured logs use token-based inputs and enables to collect
log entries from multiple sources into one destination log. This can be useful
in an autoscaling environment. You can reuse the same configuration file
multiple times without creating new hosts.

Each log to follow has a separate section in the configuration of the form:

	[name]
	path = /path/to/log/file
	token = MY_TOKEN

Where:

-  *name* is an identifier of the application that is added to your log entries
-  *path* is an absolute path to the file you wish to follow
-  *token* is the token for destination log created in Logentries

Alternatively, instead of `token` specify `destination` parameter in the format
of `host name/log name'. The agent will search for the host and log identified
by their name and retrieve the token automatically. If the host or log does not
exist, it is created.

**Note**: When using the destination parameter it is advised not to initialize
multiple agents with the same configuration file at the same time. This is to
prevent a race condition where duplicate Log Sets may be created.

Example:

	[name]
	path = /path/to/log/file
	destination = MyHost/MyLog


Using local configuration only
------------------------------

In an auto scaling environment you may not want to create a Host (log set) each
time you install the agent.

To disable pulling server-side configuration (and thus avoiding communication
with Logentries API) add this line in the `[Main]` section of the configuration:

	pull-server-side-config=False

Or specify `--pull-server-side-config=False` on the command like for the `init`
or `reinit` commands:

	sudo le reinit --pull-server-side-config=False

When set to False, the agent requires token for all configured logs. Tokens may
be specified in the configuration file, or--in case of logs created by the
agent--it is stored in local cache file.

By default, locally configured logs are sent to Logentries in Syslog format RFC
5424 which prepends a timestamp and other useful information. If you wish to
disable this, you can set the formatter to 'plain' in the `[Main]` section of
the configuration.

	formatter = plain


List IP addresses the agent uses
--------------------------------

Run the `ls ips` command to get a list of IP addresses the agent uses. These IP
addresses needs to be whitelisted in firewall.


Follow logs that change their names
--------------------------------------

Certain log rollover policies do not allow to specify destination file name.
That is typical when log files are timestamped or assigned a sequential number.
The Logentries agent can handle there cases for you.  The Logentries agent can
be pointed at particular folders to gather any active logs from that directory
or its subdirectories using wildcards in file names.  For example, the
following patterns can be used with the follow command to gather logs from the
given directories:

	/var/log/mysystem/mylog-*.log

Using wildcards when specifying the log to follow allows for situations where
you need to follow the most recent log in a particular folder. The agent looks
for any active log in the folder and will monitor the events in that log.

**Note** wildcards are NOT needed in typical syslog log rotation scheme, where
log file named `xxx.log` is renamed to `xxx.log.0` and a new `xxx.log` file is
created. In this situation follow the `xxx.log` file only, do not specify
wildcards.


Follow a log that appears across multiple directories
-----------------------------------------------------
When using the 'follow' command to set up which log file to be followed, a special
wildcard with restricted behaviour can be used for (or as part of) a directory
name. This option is intended for where multiple log files, with the same filename
but with each log file in a different directory, are to be followed. The log
events captured in all these log files are then sent to a single destination log in
the Logentries infrastructure. Note that this behaviour is different from where
a wildcard is used in the log filename - which is not allowed in this case.

Example:
For this example there is a number of directories in /var/log/, each called 'apache-xx',
where the xx is a number that varies for each directory, such as 'apache-01',
'apache-02' and so on. In addition, within each directory, is a file called 'apache.log'.

We want to follow all 'apache.log' files in each of these directories, with the log
events from all log files sent to a destination log in the Logentries infrastructure
that we want to name 'Apache'.

The command 'follow' is combined with the '--multilog' option to use the wildcard
in the directory name in the path that is passed to the agent.

    le follow '/var/log/apache-*/apache.log' --multilog [--name Apache]

#### Usage
The agent behaviour when usage of the --multilog is made is dependant on
the rules below:

-   The pathname to be passed to the agent must be in single quotes as shown
    in the example. This is necessary, so that the wildcard is not expanded
    by the shell, but processed by the agent internally.
-   The wildcard * is the only one allowed for this behaviour.
-   Only one wildcard is allowed in the pathname.
-   The wildcard can be used in place of a directory name or in place of part
    of a directory name.
-   At all times a full filename must be given.
-   No wildcard is allowed in the filename.
-   Only the one file in each directory found that matches the pattern with
    the wildcard is followed.
-   The wildcard is not required, but then only a single file is followed.
    In such a case, the --multilog option is unnecessary as the default
    behaviour of the 'follow' command can be used to follow a single file.
-   Where the --name is not used, the filename is used to name the destination
    log - which in the case of the example above is 'apache.log'
-   The agent expects the last / as being followed by a filename - it cannot
    be blank - the log filename does not need to have an extension such
    as .log; a log file named 'apache', such as '/var/log/apache*/apache'
    is accepted - as long as it is a full and valid filename.

#### Displayed Lists

If the pathname with the wildcard is valid and accepted by the agent, it will
display to the user two lists that are shown as an aid to help the user
make the most of the wildcard usage.

The first list will display all the existing destination logs for that host found
in use from the server with the pathname of the local log file (or files) associated
with each log. Where there is a log with a pathname that was set up using the --multilog
option a prefix to the pathname 'Multiple:' is shown. This list may show a mix of pathnames -
as some will have been set-up using the --mulitlog option and some not.

So as per the example above, for the log 'Apache', may return a list like this:

    LOG         PATHNAME
    mylog       /var/log/mysystem/mylog-*.log
    Apache      Multiple:/var/log/apache*/apache.log
    Cassandra   /srv/log/cassandra/system.out

The second list will display any files found by the agent at this time after
processing the pathname that fall within the scope of the wildcard. This list
reflects the files that will be followed by the agent for this pathname however,
the files actually followed may differ if directories are deleted or added
before the agent has been restarted or is actively monitoring the log files.

The agent will detect where directories that have a file being followed
is deleted, and will stop following that file. Like wise the agent will monitor
for directories with a file that falls within the scope of the wildcard expansion.
If such a directory is created while the agent is active, it will attempt to
follow this new file - with the same log settings.

The user then has the option to quit without setting any new files to be followed
or the user may continue, in which case the agent will process the pathname.

**Note** that there is no restriction on different destination logs having pathnames
with wildcards that are similar. It is possible therefore that a number of files
will appear to be common to these destination logs. The agent will not prevent this.
It is dependant on the user to understand the scope of the wildcard in the pathname
that is being applied.

For example the following two examples result in a subset of directories with the
same log file falling within the scope of both destination directories:

    le follow '/var/log/apache*/apache.log' --multilog --name=Apace
    le follow '/var/log/*/apache.log' --multilog --name=WebLogs

It is recommended to avoid scenarios like this.

### Client Configuration File
As explained above in the section on using the client side configuration to
follow files, a similar configuration section can used to set up a pathname
with a wildcard - as long as the pathname meets the same rules for one
entered via the command line.

The prefix Multilog: is mandatory.

Example:

    [Apache]
	path = Multilog:/var/log/apache*/current
	token = MY_TOKEN


Save state of file positions
----------------------------

When started in the monitoring mode, the agent follows logs from their current
position. Infrequently, in cases which requires restarting the agent, log
entries logged in the stop-start time period are not captured.


debian/changelog
logentries (1.4.43) wheezy jessie lucid precise quantal saucy trusty; urgency=low

  [ Gerald Curran ]
  * Update README.md

  [ Robin Jacobs ]
  * LOG-11043 Removed structures
  * LOG-11043 Fix distro/version check
  * LOG-11043 Fixed multilog_pathname_verification tests
  * LOG-11043 Removed unused argument from cmd_ls_ips
  * LOG-10722 Retry on 504
  * LOG-10722 Add extra information on request failure
  * LOG-10722 Sanitise log output

  [ Logentries Team ]

 -- Logentries Team <hello@logentries.com>  Wed, 13 Jun 2018 10:03:24 +0000

logentries (1.4.42) wheezy jessie lucid precise quantal saucy trusty; urgency=low

  [ Viliam Holub ]
  * Fixing error reporting
  * Adding support for structure patterns
  * Polishing code
  * Improve readability and fix comment typo

  [ Emma Louise Lynch ]
  * Docker (#206)

  [ Logentries Team ]

 -- Logentries Team <hello@logentries.com>  Tue, 30 Jan 2018 16:45:44 +0000

logentries (1.4.41) wheezy jessie lucid precise quantal saucy trusty; urgency=low

  [ Stephen Hynes ]
  * Update mac install instructions

  [ Viliam Holub ]
  * Remove guest and guest_nice from total cpu time
  * Accept application token for process metrics
  * Handle termination conditions gracefully; improve log messages for sections that do not contain path and token parameters
  * Fix spelling in pidfile error message
  * Allow process-only metrics
  * Avoid changing process title
  * Avoid race condition in tests

  [ Logentries Team ]

 -- Logentries Team <hello@logentries.com>  Thu, 19 Jan 2017 15:10:49 +0000

logentries (1.4.40) wheezy jessie lucid precise quantal saucy trusty; urgency=low

  [ Stephen Hynes ]
  * Fix for 16452

  [ Logentries Team ]

 -- Logentries Team <hello@logentries.com>  Thu, 15 Sep 2016 16:04:17 +0000

logentries (1.4.39) wheezy jessie lucid precise quantal saucy trusty; urgency=low

  [ Viliam Holub ]
  * Avoid close cycle if state file is not specified
  * Sync with currently supported systems

  [ Logentries Team ]

 -- Logentries Team <hello@logentries.com>  Thu, 15 Sep 2016 08:52:58 +0000

logentries (1.4.38) wheezy jessie lucid precise quantal saucy trusty; urgency=low

  [ Viliam Holub ]
  * Info about SHA256 digest in Debian and Ubuntu
  * Use long GPG keys
  * Save file positions
  * Fix template declaration
  * Fix obsolete V1-metrics message

  [ Stephen Hynes ]
  * Remove v1 stats api

  [ Logentries Team ]

 -- Logentries Team <hello@logentries.com>  Wed, 14 Sep 2016 16:42:29 +0000

logentries (1.4.37) wheezy jessie lucid precise quantal saucy trusty; urgency=low

  [ Viliam Holub ]
  * Add instructions for Ubuntu 16.04 installation

  [ Logentries Team ]

 -- Logentries Team <hello@logentries.com>  Fri, 20 May 2016 15:56:13 +0000

logentries (1.4.36) wheezy jessie lucid precise quantal saucy trusty; urgency=low

  [ Viliam Holub ]
  * Fix trailing space in help.sh test

  [ Logentries Team ]

 -- Logentries Team <hello@logentries.com>  Fri, 13 May 2016 16:00:14 +0000

logentries (1.4.35) wheezy jessie lucid precise quantal saucy trusty; urgency=low

  [ Mr Jenkins ]
  * Release new version 1.4.34

  [ Viliam Holub ]
  * Send IAA token via transport layer

  [ Logentries Team ]

 -- Logentries Team <hello@logentries.com>  Fri, 13 May 2016 15:11:13 +0000

logentries (1.4.34) wheezy jessie lucid precise quantal saucy trusty; urgency=low

  [ Charles Phillips ]
  * Add socks.py from #165 to fix ImportError line 264

  [ Paul CLune ]
  * Log-7549 and LOG-7550 Linux Agent Stories
  * LOG-7549 Styling issues addressed

  [ Logentries Team ]

 -- Logentries Team <hello@logentries.com>  Thu, 05 May 2016 15:22:25 +0000

logentries (1.4.33) wheezy jessie lucid precise quantal saucy trusty; urgency=low

  [ Viliam Holub ]
  * Resolve embedded structure names
  * Set Amazon\'s repository to a fixed address
  * FTR-59 Remove explicit hostname and appname from default format
  * Use apt-key to get GPG keys
  * Read log from the beginning when it shrinks

  [ Marc Simon ]
  * [improv] using proxy for http request

  [ Jon Tibble ]
  * Cleanup directory creation assumptions

  [ Logentries Team ]

 -- Logentries Team <hello@logentries.com>  Mon, 21 Mar 2016 15:10:18 +0000

logentries (1.4.32) wheezy jessie lucid precise quantal saucy trusty; urgency=low

  [ Ulrich Dangel ]
  * Fix comparison against server names

  [ Aurelio Jargas ]
  * Fix indentation

  [ Viliam Holub ]
  * Avoid cache when server side configuration is enabled
  * Removing local cache

  [ Logentries Team ]

 -- Logentries Team <hello@logentries.com>  Wed, 24 Feb 2016 19:12:58 +0000

logentries (1.4.31) wheezy jessie lucid precise quantal saucy trusty; urgency=low

  [ Ulrich Dangel ]
  * Fix tests by only trying to change file which exist

  [ Logentries Team ]

 -- Logentries Team <hello@logentries.com>  Thu, 04 Feb 2016 17:38:33 +0000

logentries (1.4.30) wheezy jessie lucid precise quantal saucy trusty; urgency=low

  [ Viliam Holub ]
  * Improving error response when host is not found
  * Improving "File does not exist" warning

  [ Ulrich Dangel ]
  * Adjust configuration file permissions

  [ Logentries Team ]

 -- Logentries Team <hello@logentries.com>  Thu, 04 Feb 2016 17:05:57 +0000

logentries (1.4.29) wheezy jessie lucid precise quantal saucy trusty; urgency=low

  [ Viliam Holub ]
  * Adding support for CentOS7
  * Improving style, adding instructions for FreeBSD 10.2
  * Not dumping stact trace for Ctrl+D in get_pass
  * Updating internal list of root CAs

  [ Logentries Team ]

 -- Logentries Team <hello@logentries.com>  Tue, 12 Jan 2016 11:05:26 +0000

logentries (1.4.28) wheezy jessie lucid precise quantal saucy trusty; urgency=low

  [ Charles Phillips ]
  * Update install to latest Logentries and SmartOS base image
  * v1 metrics for SmartOS [#60]

  [ Jimmy Tang ]
  * Initial add of reference homebrew packaging.
  * create a small script for generating an updated formula

  [ Viliam Holub ]
  * Adding instructions for Ubuntu Wily Werewolf


src/utils.py

# coding: utf-8
# vim: set ts=4 sw=4 et:

import errno
import httplib
import os
import re
import socket
import sys
try:
    import uuid
    FEAT_UUID = True
except ImportError:
    FEAT_UUID = False
from backports import match_hostname, CertificateError

import logging


__author__ = 'Logentries'

__all__ = ["EXIT_OK", "EXIT_NO", "EXIT_HELP", "EXIT_ERR", "EXIT_TERMINATED",
           "ServerHTTPSConnection", "LOG_LE_AGENT", "create_conf_dir",
           "default_cert_file", "system_cert_file", "domain_connect",
           "no_more_args", "find_hosts", "find_logs", "find_api_obj_by_key", "find_api_obj_by_name",
           "die", "error", "colored", "red", "rfile", 'TCP_TIMEOUT', "rm_pidfile", "uuid_parse",
           "report"]

# Return codes
EXIT_OK = 0
EXIT_NO = 1
EXIT_ERR = 3
EXIT_HELP = 4
EXIT_TERMINATED = 5  # Terminated by user (Ctrl+C)

LE_CERT_NAME = 'ca-certs.pem'

TCP_TIMEOUT = 10  # TCP timeout for the socket in seconds


authority_certificate_files = [  # Debian 5.x, 6.x, 7.x, Ubuntu 9.10, 10.4, 13.0
                                 "/etc/ssl/certs/ca-certificates.crt",
                                 # Fedora 12, Fedora 13, CentOS 5
                                 "/usr/share/purple/ca-certs/GeoTrust_Global_CA.pem",
                                 # Amazon AMI, CentOS 7, recent RHs
                                 "/etc/pki/tls/certs/ca-bundle.crt",
                                 # FreeBSD 10.2
                                 "/etc/ssl/cert.pem",
]

LOG_LE_AGENT = 'logentries.com'

log = logging.getLogger(LOG_LE_AGENT)

try:
    import ssl

    wrap_socket = ssl.wrap_socket
    FEAT_SSL = True
    try:
        ssl.create_default_context
        FEAT_SSL_CONTEXT = True
    except AttributeError:
        FEAT_SSL_CONTEXT = False
except ImportError:
    FEAT_SSL = False
    FEAT_SSL_CONTEXT = False

    def wrap_socket(sock, ca_certs=None, cert_reqs=None):
        return socket.ssl(sock)

def report(what):
    print >> sys.stderr, what

class ServerHTTPSConnection(httplib.HTTPSConnection):

    """
    A slight modification of HTTPSConnection to verify the certificate
    """

    def __init__(self, config, server, port, cert_file):
        self.no_ssl = config.suppress_ssl or not FEAT_SSL
        if self.no_ssl:
            if config.use_proxy == True:
                httplib.HTTPSConnection.__init__(self, config.proxy_url, config.proxy_port, context=context)
                if hasattr(httplib.HTTPSConnection, "set_tunnel"):
                    httplib.HTTPSConnection.set_tunnel(self, server, port)
                else:
                    httplib.HTTPSConnection._set_tunnel(self, server, port)
            else:
                httplib.HTTPSConnection.__init__(self, server, port)
        else:
            self.cert_file = cert_file
            if FEAT_SSL_CONTEXT:
                context = ssl.create_default_context(cafile=cert_file)
                if config.use_proxy == True:
                    httplib.HTTPSConnection.__init__(self, config.proxy_url, config.proxy_port, context=context)
                    if hasattr(httplib.HTTPSConnection, "set_tunnel"):
                        httplib.HTTPSConnection.set_tunnel(self, server, port)
                    else:
                        httplib.HTTPSConnection._set_tunnel(self, server, port)
                else:
                    httplib.HTTPSConnection.__init__(self, server, port, context=context)
            else:
                if config.use_proxy == True:
                    httplib.HTTPSConnection.__init__(self, config.proxy_url, config.proxy_port, cert_file=cert_file)
                    if hasattr(httplib.HTTPSConnection, "set_tunnel"):
                        httplib.HTTPSConnection.set_tunnel(self, server, port)
                    else:
                        httplib.HTTPSConnection._set_tunnel(self, server, port)
                else:
                    httplib.HTTPSConnection.__init__(self, server, port, cert_file=cert_file)

    def connect(self):
        if FEAT_SSL_CONTEXT:
            httplib.HTTPSConnection.connect(self)
        else:
            if self.no_ssl:
                return httplib.HTTPSConnection.connect(self)
            sock = create_connection(self.host, self.port)
            try:
                if self._tunnel_host:
                    self.sock = sock
                    self._tunnel()
            except AttributeError:
                pass
            self.sock = wrap_socket(
                sock, ca_certs=self.cert_file, cert_reqs=ssl.CERT_REQUIRED)
            try:
                match_hostname(self.sock.getpeercert(), self.host)
            except CertificateError, ce:
                die("Could not validate SSL certificate for %s: %s" % (
                    self.host, ce.message))


def default_cert_file_name(config):
    """
    Construct full file name to the default certificate file.
    """
    return config.config_dir_name + LE_CERT_NAME


def create_conf_dir(config):
    """
    Creates directory for the configuration file.
    """
    # Create logentries config
    try:
        os.makedirs(config.config_dir_name)
    except OSError, e:
        if e.errno != errno.EEXIST:
            if e.errno == errno.EACCES:
                die("You don't have permission to create logentries config file. Please run logentries agent as root.")
            die('Error: %s' % e)


def write_default_cert_file(config, authority_certificate):
    """
    Writes default certificate file in the configuration directory.
    """
    create_conf_dir(config)
    cert_filename = default_cert_file_name(config)
    f = open(cert_filename, 'wb')
    f.write(get_bundled_certs())
    f.close()


def default_cert_file(config):
    """
    Returns location of the default certificate file or None. It tries to write the
    certificate file if it is not there or it is outdated.
    """
    cert_filename = default_cert_file_name(config)
    try:
        # If the certificate file is not there, create it
        if not os.path.exists(cert_filename):
            write_default_cert_file(config, authority_certificate)
            return cert_filename

        # If it is there, check if it is outdated
        curr_cert = rfile(cert_filename)
        if curr_cert != authority_certificate:
            write_default_cert_file(config, authority_certificate)
    except IOError:
        # Cannot read/write certificate file, ignore
        return None
    return cert_filename


def system_cert_file():
    """
    Finds the location of our lovely site's certificate on the system or None.
    """
    for f in authority_certificate_files:
        if os.path.exists(f):
            return f
    return None


def create_connection(host, port):
    """
    A simplified version of socket.create_connection from Python 2.6.
    """
    for addr_info in socket.getaddrinfo(host, port, 0, socket.SOCK_STREAM):
        af, stype, proto, cn, sa = addr_info
        soc = None
        try:
            soc = socket.socket(af, stype, proto)
            soc.settimeout(TCP_TIMEOUT)
            soc.connect(sa)
            return soc
        except socket.error:
            if socket:
                soc.close()

    raise socket.error, "Cannot make connection to %s:%s" % (host, port)


def make_https_connection(config, s, port):
    """
    Makes HTTPS connection. Tried all available certificates.
    """
    if not config.use_ca_provided:
        # Try to connect with system certificate
        try:
            cert_file = system_cert_file()
            if cert_file:
                return ServerHTTPSConnection(config, s, port, cert_file)
        except socket.error, e:
            pass

    # Try to connect with our default certificate
    cert_file = default_cert_file(config)
    if not cert_file:
        die('Error: Cannot find suitable CA certificate.')
    return ServerHTTPSConnection(config, s, port, cert_file)


def domain_connect(config, domain, Domain):
    """
    Connects to the domain specified.
    """
    # Find the correct server address
    s = domain
    if Domain == Domain.API:
        if config.force_domain:
            s = config.force_domain
        elif config.force_api_host:
            s = config.force_api_host
        else:
            s = Domain.API

    # Special case for local debugging
    if config.debug_local:
        if s == Domain.API:
            s = Domain.API_LOCAL
        else:
            s = Domain.MAIN_LOCAL

    # Determine if to use SSL for connection
    # Never use SSL for debugging, always use SSL with main server
    use_ssl = True
    if config.debug_local:
        use_ssl = False
    elif s == Domain.API:
        use_ssl = not config.suppress_ssl

    # Connect to server with SSL in untrusted network
    if use_ssl:
        port = 443
    else:
        port = 80
    if config.debug_local:
        if s == Domain.API:
            port = 8000
        else:
            port = 8081
    log.debug('Connecting to %s:%s', s, port)

    # Pass the connection
    if use_ssl:
        return make_https_connection(config, s, port)
    else:
        if config.use_proxy == True:
            conn = httplib.HTTPConnection(config.proxy_url, config.proxy_port)
            if hasattr(httplib.HTTPConnection, "set_tunnel"):
                conn.set_tunnel(s, port)
            else:
                conn._set_tunnel(s, port)
            return conn
        else:
            return httplib.HTTPConnection(s, port)



def no_more_args(args):
    """
    Exits if there are any arguments given.
    """
    if len(args) != 0:
        die("No more than one argument is expected.")


def expr_match(expr, text):
    """
    Returns True if the text matches with expression. If the expression
    starts with / it is a regular expression.
    """
    if expr[0] == '/':
        if re.match(expr[1:], text):
            return True
    else:
        if expr[0:2] == '\\/':
            return text == expr[1:]
        else:
            return text == expr
    return False


def find_hosts(expr, hosts):
    """
    Finds host name among hosts.
    """
    result = []
    for host in hosts:
        if uuid_match(expr, host['key']) or expr_match(expr, host['name']) or expr_match(expr, host['hostname']):
            result.append(host)
    return result


def log_match(expr, log_item):
    """
    Returns true if the expression given matches the log. Expression is either
    a simple word or a regular expression if it starts with '/'.

    We perform the test on UUID, log name, and file name.
    """
    return uuid_match(
        expr, log_item['key']) or expr_match(expr, log_item['name']) or expr_match(expr,
                                                                                   log_item['filename'])


def find_logs(expr, hosts):
    """
    Finds log name among hosts. The searching expresion have to parts: host
    name and logs name. Both parts are divided by :.
    """
    # Decode expression
    l = expr.find(':')
    if l != -1:
        host_expr = expr[0:l]
        log_expr = expr[l + 1:]
    else:
        host_expr = '/.*'
        log_expr = expr

    adepts = find_hosts(host_expr, hosts)
    logs = []
    for host in adepts:
        for xlog in host['logs']:
            if log_match(log_expr, xlog):
                logs.append(xlog)
    return logs


def find_api_obj_by_name(obj_list, name):
    """
    Finds object in a list by its name parameter. List of objects must conform
    to that of a log or host entity from api.
    """
    result = None
    for obj in obj_list:
        if obj['name'] == name:
            result = obj
            break
    return result


def find_api_obj_by_key(obj_list, key):
    """
    Finds object in a list by its key parameter. List of objects must conform
    to that of a log or host entity from api.
    """
    result = None
    for obj in obj_list:
        if obj['key'] == key:
            result = obj
            break
    return result


def die(cause, exit_code=EXIT_ERR):
    log.critical('\n%s', cause)
    sys.exit(exit_code)


def rfile(name):
    """
    Returns content of the file, without trailing newline.
    """
    x = open(name).read()
    if len(x) != 0 and x[-1] == '\n':
        x = x[0:len(x) - 1]
    return x


def rm_pidfile(config):
    """
    Removes PID file. Called when the agent exits.
    """
    try:
        if config.pid_file:
            os.remove(config.pid_file)
    except OSError:
        pass


def uuid_match(uuid, text):
    """
    Returns True if the uuid given is uuid and it matches to the text.
    """
    return is_uuid(uuid) and uuid == text


def is_uuid(x):
    """
    Returns true if the string given appears to be UUID.
    """
    return re.match(r'[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}', x)


def uuid_parse(text):
    """Returns uuid given or None in case of syntax error.
    """
    try:
        if FEAT_UUID:
            return uuid.UUID(text).__str__()
        else:
            low_text = text.lower()
            if re.match( r'^[a-f0-9]{8}-[a-f0-9]{4}-4[a-f0-9]{3}-[89ab][a-f0-9]{3}-[a-f0-9]{12}', low_text):
                return low_text
    except ValueError:
        pass
    return None

authority_certificate = ""


def get_bundled_certs():
    file_name = os.path.join(os.path.dirname(__file__), "cacert.pem")
    with open(file_name) as r:
        return r.read()

try:
    import termcolor
    colored = termcolor.colored
except ImportError:
    def colored(text, color):
        return text

def red(text):
    return colored('%s'%text, 'red')

def error(cause, *args):
    die(red('Error:' + ' ' + cause%args))
